File: c/src/ValidAnagram.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

static bool isAnagram(char *, char *);
static void print(char *, char *, bool);
static void testplusplus();

int main() {
    /*
    char *s1 = "anagram";
    char *t1 = "nagaram";
    char *s2 = "rat";
    char *t2 = "car";
    bool r1 = isAnagram(s1,t1);
    bool r2 = isAnagram(s2,t2);
    print(s1,t1,r1);
    print(s2,t2,r2);
    */
   testplusplus();
}

static bool isAnagram(char *s, char *t) {
    int len = strlen(s);
    if (strlen(t) != len) { return false; }
    int freq[26] = {0}; // the local array will not be initialized to 0 if I declare by with "freq[26];"
    for (int i = 0; i < len; i++) {
        freq[*s++ - 'a']++;
        freq[*t++ - 'a']--;
    }
    for (int i = 0; i < 26; i++) {
        if (freq[i]) { return false; }
    }
    return true;
}

static void print(char *s, char *t, bool result) {
    printf("[%s] and [%s] are anagram? %s\n", s, t, result? "true" : "false");
}

static void testplusplus() {
    int nums[10] = {0};
    for (int i = 0; i < 10; ++i) {
        nums[i] = i;
    }
    printf("[");
    for (int i = 0; i < 10; i++) {
        printf("%d",nums[i]);
        if (i < 9) { printf(","); }
    }
    printf("]\n");
}

File: java/src/main/java/com/ciaoshen/leetcode/ValidAnagram.java
/**
 * Leetcode - Algorithm - Valid Anagram
 */
package com.ciaoshen.leetcode;
import java.util.*;

class ValidAnagram {
    public class SolutionV1 {
        private final int[] CODE = new int[]{2,3,5,7,11,13,17,19,23,29,31,37,41,43,47,53,59,61,67,71,73,83,89,93,97,101};
        public boolean isAnagram(String s, String t) {
            return encode(s) == encode(t);
        }
        public long encode(String s) {
            long r = 1;
            for (int i = 0; i < s.length(); i++) {
                r *= CODE[s.charAt(i) - 'a'];
            }
            return r;
        }
    }
    public class SolutionV2 {
        public boolean isAnagram(String s, String t) {
            Map<Character,Integer> freq = new HashMap<>();
            int len = s.length();
            if (t.length() != len) { return false; }
            for (int i = 0; i < 26; i++) {
                freq.put((char)(i+'a'),0);
            }
            for (int i = 0; i < len; i++) {
                Character c = s.charAt(i);
                freq.put(c,freq.get(c)+1);
            }
            for (int i = 0; i < len; i++) {
                Character c = t.charAt(i);
                freq.put(c,freq.get(c)-1);
            }
            for (Map.Entry<Character,Integer> entry : freq.entrySet()) {
                if (entry.getValue() != 0) { return false; }
            }
            return true;
        }
    }
    public class Solution {
        public boolean isAnagram(String s, String t) {
            int len = s.length();
            if (t.length() != len) { return false; }
            int[] freq = new int[26];
            for (int i = 0; i < len; i++) {
                freq[s.charAt(i)-'a']++;
                freq[t.charAt(i)-'a']--;
            }
            for (int i = 0; i < 26; i++) {
                if (freq[i] != 0) { return false; }
            }
            return true;
        }
    }
    private class Test {
        private void callIsAnagram(String s, String t) {
            boolean result = solution.isAnagram(s,t);
            String keyWord = (result)? " are " : " are not ";
            System.out.println("[" + s + "]" + " and " + "[" + t + "]" + keyWord + "anagram!");
        }
        private void test() {
            String s1 = "anagram";
            String t1 = "nagaram";
            String s2 = "rat";
            String t2 = "car";
            // String s3 =;
            // String t3 =;
            callIsAnagram(s1,t1);
            callIsAnagram(s2,t2);
            // callIsAnagram(s3,t3);
        }
    }
    private static ValidAnagram problem = new ValidAnagram();
    private static Solution solution = problem.new Solution();
    private static Test test = problem.new Test();
    public static void main(String[] args) {
        test.test();
    }
}
